embrace-android-features/src/main/kotlin/io/embrace/android/embracesdk/internal/capture/activity/UiLoadEventEmitter.kt

```diff
@@ -0,0 +1,149 @@
+package io.embrace.android.embracesdk.internal.capture.activity
+
+import android.app.Activity
+import android.app.Application.ActivityLifecycleCallbacks
+import android.os.Build
+import android.os.Bundle
+import io.embrace.android.embracesdk.internal.clock.nanosToMillis
+import io.embrace.android.embracesdk.internal.session.lifecycle.ActivityLifecycleListener
+import io.embrace.android.embracesdk.internal.utils.VersionChecker
+import io.opentelemetry.sdk.common.Clock
+
+/**
+ * Maps [ActivityLifecycleCallbacks] events to [UiLoadEvents] depending on the current state of the app and capabilities of the OS.
+ *
+ * The purpose of this is to leverage Activity lifecycle events to provide data for the underlying workflow to bring a new Activity on
+ * screen. Due to the varying capabilities of the APIs available on the different versions of Android, the precise triggering events for
+ * the start and intermediate steps may differ.
+ *
+ * See [UiLoadTraceEmitter] for details.
+ */
+class UiLoadEventEmitter(
+    private val uiLoadEvents: UiLoadEvents,
+    private val clock: Clock,
+    private val versionChecker: VersionChecker,
+) : ActivityLifecycleListener {
+
+    override fun onActivityPreCreated(activity: Activity, savedInstanceState: Bundle?) {
+        create(activity)
+    }
+
+    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
+        if (!versionChecker.firePrePostEvents()) {
+            create(activity)
+        }
+    }
+
+    override fun onActivityPostCreated(activity: Activity, savedInstanceState: Bundle?) {
+        createEnd(activity)
+    }
+
+    override fun onActivityPreStarted(activity: Activity) {
+        start(activity)
+    }
+
+    override fun onActivityStarted(activity: Activity) {
+        if (!versionChecker.firePrePostEvents()) {
+            createEnd(activity)
+            start(activity)
```

Comment:
I find the naming of these functions tricky to follow. I think readability would be improved by naming them `onActivityStarted` or something similar. Alternatively there might be scope for representing the potential states with an enum/sealed class hierarchy and routing events through one function?

Reply:
So these get moved around in [this PR](https://github.com/embrace-io/embrace-android-sdk/pull/1683)

While the names haven't change, not mixing up two implementations in one make them a little more clear. 

And this was a half-hearted effort to decouple the UILoad events from the activity lifecycle events - these stages SHOULD be generic so we can support different view loading lifecycles (e.g. compose), so I've made it possible to genericize it without actually doing all the work right now (because it's not needed).

For stacking an merging reasons, I'd like to merge this as is, and we can look at #1683 to see if there's a better way of naming these things, or maybe even doing the genericization there?